[WIP] Fixed issue #17: Bookmarks

antoyo opened the pull request:
Hi.
This is the start of my work for the issue #17 (bookmarks).
With the current work, it is possible to save and open bookmarks (with the completion).
It is missing the completion on the name (currently, it only works on url) and the feature to delete bookmarks.
I'm creating this pull request right now because I want to know what to do for these missing features.
Is there anything similar to the bookmark removal feature (can we delete quickmark or anything else in qutebrowser)?
Second, I don't know where to update the code to support the bookmark completion by title. Can you point me where to look at, please?

The-Compiler commented:
Thanks for you first contribution! :smile: I don't really get what the difference of this compared to quickmarks is - could you explain?

I gave it a quick look and try. Code-wise it seems it's mostly the same as the quickmarks code, or am I missing something?

It seems quickmark completion for `:open` is removed (is this intentional?) and `:bookmark-add` crashes because `url` is a string, not a `QUrl`.

I imagined bookmarks more like an addition to the existing quickmarks feature - as mentioned in the issue, perhaps just quickmarks with no name, or an autogenerated one.

antoyo commented:
As far as I understand, quickmarks maps a name to a URL. I do not really use quickmarks, so I don't know if there is any other difference with bookmarks.
In dwb, there are both quickmarks and bookmarks, so I believe there is room for both of them in qutebrowser.
As far as I know, with bookmarks, we would have auto-completion for both the URL and the page title.
Yes, the code is very similar to quickmarks.
You're right, the quickmark completion is now broken and `:bookmark-add` crashes: I will fix these issues.
But, perhaps we could add URL completion on quickmarks and have the possibility to use the page title for the name and combine bookmarks with quickmarks.
What do you prefer?

The-Compiler commented:
> As far as I understand, quickmarks maps a name to a URL. I do not really use quickmarks, so I don't know if there is any other difference with bookmarks.

Yeah - but that's the same thing your bookmarks do as well now, or isn't it?

> In dwb, there are both quickmarks and bookmarks, so I believe there is room for both of them in qutebrowser.

Hmm - I'll have to check how it's handled there and what the differences are.

> As far as I know, with bookmarks, we would have auto-completion for both the URL and the page title.

quickmarks currently have completion for the quickmark name (when using <kbd>b</kbd>/`:quickmark-load`) or the URL (when using <kbd>o</kbd>/`:open`). Adding the page title for `:open` is probably not something trivial I'm afraid, as we can currently just use Qt's built-in filtering which is much faster than custom filtering in Python.

> Yes, the code is very similar to quickmarks.

Then maybe it'd make more sense to simply add bookmarks functionality to the existing code (i.e. what's now called quickmarks) instead of duplicating much code.

But first we need to know what the differences between quickmarks and bookmarks actually should be.

Here's what I think, without researching how others differentiate bookmarks/quickmarks:

* Bookmarks don't have any kind of mandatory label
* Bookmarks have more features (can maybe be organized in a folder hierarchy)

Maybe a good start would be to add a `~/.config/qutebrowser/bookmarks` folder, and then read the file `~/.config/qutebrowser/bookmarks/urls` (so folders can be added later, and are just sub-folders in the file structure). That file would just be a list of URLs for now.

Then completion support for that could be added (for the URL only for now) for `:open`, as well as new commands to add a new bookmark for the current or a given page (similiar to quickmarks, but *without* name).

What do you think about this? Then we can always go from there.

> But, perhaps we could add URL completion on quickmarks and have the possibility to use the page title for the name and combine bookmarks with quickmarks.

That sounds better, yes. But I think getting the page title and keeping it up to date is non-trivial, so maybe leave that for later.

The-Compiler commented:
Here is the log of what we discussed in IRC so this doesn't get lost:

```
16:28 <antoyo> Does having a command to add a quickmark with the current page title as the name would be ok?
16:32 <The-Compiler> I wondered about that too... But I think they shouldn't be in the quickmarks file then :-/
16:32 <antoyo> Why?
16:33 <The-Compiler> Because they aren't quickmarks - for example when using bookmarks for things I want to look at later or just "find cool", I don't want them to be in the completion with b/:quickmark-load
16:35 <antoyo> It seems I use bookmarks the same way you use quickmarks. Perhaps we need to know the exact differences between them. Unfortunately, I cannot run dwb anymore since it crashes since an update that happened yesterday.
16:38 <The-Compiler> dwb seems to have :bookmark which bookmarks the current page and then a :bookmarks menu to retrieve them
16:39 <antoyo> What about the completion? I know dwb complete bookmarks in :open, but does it complete quickmarks as well?
16:39 <The-Compiler> It saves "label URL title" for quickmarks (in ~/.config/dwb/quickmarks) and "URL title" for bookmarks (in ~/.config/dwb/bookmarks)   
16:40 <The-Compiler> Only bookmarks for :open - there I'd prefer both bookmarks/quickmarks, but in separate categories
16:40 <antoyo> So far, the only difference is that quickmarks have a label?
16:42 <The-Compiler> (and that they're in an own file, and own/no completion)
16:45 <antoyo> As far as I know, there is no directory structure in dwb bookmarks.
16:46 <The-Compiler> Seems like, yeah. That still would be an useful feature though. ;)
16:46 <The-Compiler> (btw, if other people are reading this, your input is appreciated!)
16:46 <antoyo> I don't see how it would be shown in the interface, though.
16:47 <antoyo> How would one go from one directory to another using the keyboard?
16:48 <The-Compiler> Right now maybe just as a slashed path - in the future, it could be a real tree and after you complete a bookmark directory name you only see completions for items in that folder.
16:50 <antoyo> So, if I fixed the bugs you found in my pull request, what will be missing?
16:51 <The-Compiler> I couldn't really test it ;) But what I don't like looking at the code right now is all the duplication. I think it'd be better if both would be handled in a common place, as the differences are minor.
16:53 <antoyo> To test it, add these line to keys.conf:
16:53 <antoyo> bookmark-save
16:53 <antoyo>     M
16:53 <antoyo> I agree that a refactoring is necessary: I will fix this as well.
16:54 <antoyo> I also need to take into account the case when there is no page title: dwb is inserting "(null)" in this case, I think.
16:55 <antoyo> Are you able to add bookmarks with the M key binding?
16:57 <The-Compiler> No time to test right now, need to catch the bus in a few minutes :) (but will be back in 15min I hope)
16:58 <antoyo> Ok
17:15 <The-Compiler> antoyo: back - another thing I'm not sure about is how to make sure using subfolders in the future will be possible
17:16 <antoyo> Hmm
17:16 <The-Compiler> antoyo: I can think of three solutions: 1) just use slashed paths in the bookmarks file too, 2) use a yaml file, 3) like I said in the pull request, use a bookmarks folder and use the filesystem structure, with an 'urls' file in each folder
17:17 <antoyo> Is there a way to easily delete a quickmark in qutebrowser? In dwb, we can delete a bookmark with Ctrl-D in the completion popup.
17:18 <The-Compiler> antoyo: :quickmark-del only currently, but I just saw that and thought it was a good idea
17:19 <antoyo> For the folder structure, I don't know.
17:19 <antoyo> How it works for default key binding in qutebrowser? How "bookmark-save\n    M" for instance is added to keys.conf?
17:22 <The-Compiler> antoyo: see KEY_DATA in config/configdata.py
17:22 <antoyo> Ah, I see.
17:23 <antoyo> And for the completion, do you use QCompleter?
17:24 <The-Compiler> no, it's essentially a QStandardItemModel with a QSortFilterProxyModel and a QTreeView
17:26 <antoyo> I wonder if there is a way to filter multiple columns with Qt.
17:27 <The-Compiler> see completion/models/sortfilter.py for the related code
17:30 <antoyo> A way of having completion on both the URL and the title would be to put both of them in data, but it looks like a hack.
17:34 <antoyo> It seems possible to filter on multiple columns: http://doc.qt.io/qt-5/qtwidgets-itemviews-customsortfiltermodel-example.html [ http://cmpl.cc:8888/3am ]
17:35 <The-Compiler> with a custom filterAcceptsRow, sure - but that's what I'd like to avoid because of performance
17:36 <antoyo> Don't we already have a filterAcceptsRow() method?
17:36 <The-Compiler> oh damn, it's Friday
17:37 <The-Compiler> I somehow kept thinking about lessThan (which is avoided with dumb_sort if possible)
17:38 <antoyo> So filtering on title and url in filterAcceptsRow() would be ok?
17:49 <The-Compiler> antoyo: I suggest this: Add a custom_filter(...) method to BaseCompletionModel which raises NotImplementedError. Then in CompletionFilterModel.filterAcceptsRow, after doing the other filtering for descriptions, call self.srcmodel.custom_filter, catch NotImplementedError and fall back to the default implementation (like sort in set_pattern). And then implement custom_filter for UrlCompletionModel.
17:51 <antoyo> So it will work for bookmarks, quickmarks and history?
17:52 <The-Compiler> It'll have to take into account that quickmarks and history entries don't have titles yet, but once that's added, sure
```

antoyo commented:
I fixed the quickmark completion and the broken command.
However, I tried to refactor the code to remove the duplicate code, but I find it hard to do so because there are some subtle differences between quickmarks and bookmarks. Do you have an idea of how to refactor or reuse some code?
I also added filtering for url and title using the method you suggested.

Next, I'll work on bookmark removal.
Should this removal feature work with quickmarks as well?

The-Compiler commented:
Thanks!

I'll take a look at how the code could be merged later - I haven't taken a closer look yet. Maybe a common subclass for `BookmarkManager` and `QuickmarkManager`?

As for the command names, I'd prefer to call them `:bookmark-{add,del,load}` so it's similiar to quickmarks.

I think it should be highlighted as well, yes. The highlighting happens in `completion/completiondelegate.py` in `CompletionDelegate:_get_textdoc()`. I guess you can simply remove the `if index.column() == 0` part, but I haven't tried yet.

What do you mean with removal exactly? The `Ctrl-D` binding you mentioned? Ideally that'd be implemented in some general way so it works for quickmarks and maybe history as well and can be extended for others in the future. Though I don't know what the best way to implement it would be.

antoyo commented:
I've fixed the command name and added the highlighting.
I also started the code to remove bookmarks and quickmarks with `Ctrl-D`. I will need to figure a way to make the code more generic.

The-Compiler commented:
I haven't thought about it in detail yet, but I imagine it something like this:

* There's a new command, `:completion-item-del` or so (mapped to `Ctrl-D` in the config by default). It should probably be in `completionwidget.py`, but wherever it fits.
* Each model has a `delete_cur_item` method or so. It's defined in `BaseCompletionModel` where it just raises `NotImplementedError`.
* The new command tries to call `delete_cur_item` - either the model handles it (in which case there's nothing more to do), or it raises `NotImplementedError` in which case `cmdexc.CommandError` should be raised to display an error to the user.

antoyo commented:
I fixed a few issues.
Is there anything else to do besides refactoring?

The-Compiler commented:
FYI, closing/reopening this to get Travis to build it.

I don't think there's anything left so far - I hope I'll get to review it later this or next week. Sorry for the delay!

The-Compiler commented:
I think the best way to reduce code duplication would be to have a base class, from which `BookmarkManager` and `QuickmarkManager` inherit.

Could you look at this in detail? If not, I can look at it, but I'm afraid it'll probably take me some more days until I find time to do so.

Here's a little example of what I imagine:

```python
class BaseMarkManager(QObject):

    """Base class to share code between QuickmarkManager and BookmarkManager.

    Attributes:
        marks: An OrderedDict of all marks.
        _lineparser: The LineParser used for the marks, or None
                     (when qutebrowser is started with -c '').

    Signals:
        changed: Emitted when anything changed.
        added: Emitted when a new mark was added.
               arg 0: The name of the mark.
               arg 1: The URL of the mark, as string.
        removed: Emitted when an existing mark was removed.
                 arg 0: The name of the mark.
    """

    FILENAME = None
    NAME = None

    def __init__(self, parent=None):
        super().__init__(parent)

        self.marks = collections.OrderedDict()

        if standarddir.config() is None:
            self._lineparser = None
        else:
            self._lineparser = lineparser.LineParser(
                standarddir.config(), self.FILENAME, parent=self)

            for line in self._lineparser:
                if not line.strip():
                    # Ignore empty or whitespace-only lines.
                    continue
                try:
                    key, url = line.rsplit(maxsplit=1)
                except ValueError:
                    message.error(0, "Invalid {} '{}'".format(self.NAME, line))
                else:
                    self.marks[key] = url
            filename = os.path.join(standarddir.config(), self.FILENAME)
            objreg.get('save-manager').add_saveable(
                '{}-manager'.format(self.NAME), self.save, self.changed,
                filename=filename)

    def save(self):
        """Save the quickmarks to disk."""
        if self._lineparser is not None:
            self._lineparser.data = [' '.join(tpl)
                                     for tpl in self.marks.items()]
            self._lineparser.save()


class QuickmarkManager(BaseMarkManager):

    """Manager for quickmarks."""

    FILENAME = 'quickmarks'
    NAME = 'quickmark'

    def prompt_save(self, win_id, url):
        # ...

    @cmdutils.register(instance='quickmark-manager', win_id='win_id')
    def quickmark_add(self, win_id, url, name):
        # ...

    @cmdutils.register(instance='quickmark-manager', maxsplit=0,
                       completion=[usertypes.Completion.quickmark_by_name])
    def quickmark_del(self, name):
        # ...

    def get(self, name):
        # ...


class BookmarkManager(BaseMarkManager):

    """Manager for bookmarks."""

    FILENAME = 'bookmarks/url'
    NAME = 'bookmarks'

    def __init__(self, parent=None):
        """Initialize and read bookmarks."""
        if standarddir.config() is not None:
            bookmarks_directory = os.path.join(standarddir.config(),
                                               'bookmarks')
            if not os.path.isdir(bookmarks_directory):
                os.makedirs(bookmarks_directory)

        super().__init__(parent)

    def bookmark_add(self, win_id, url, title):
        # ...

    def bookmark_del(self, url):
        # ...
```

I'll also add some more comments, though I'm not sure how applicable those will be after the subclassing - but it might be good to keep them in mind.

Thanks!

The-Compiler commented on qutebrowser/browser/bookmarks.py:
```diff
@@ -0,0 +1,140 @@
+                                     for tpl in self.bookmarks.items()]
+            self._lineparser.save()
+
+    def bookmark_add(self, win_id, url, title):
```
This should probably just be called `add` - the `bookmark_` prefix is not needed (it's there for `quickmark_add` because that's registered as a command).

The-Compiler commented on qutebrowser/browser/bookmarks.py:
```diff
@@ -0,0 +1,140 @@
+            self.added.emit(title, urlstr)
+            message.info(win_id, "Bookmark added")
+
+    def bookmark_del(self, url):
```
See above for the naming.

The-Compiler commented on qutebrowser/browser/bookmarks.py:
```diff
@@ -0,0 +1,140 @@
+        try:
+            del self.bookmarks[url]
+        except KeyError:
+            raise cmdexc.CommandError("Bookmark '{}' not found!".format(url))
```
Since this isn't a command, it probably shouldn't raise a `cmdexc.CommandError`. The `KeyError` here should never happen as it's only called via the completion anyways, right? Then I'd say remove the whole try/except-block.

The-Compiler commented on qutebrowser/browser/bookmarks.py:
```diff
@@ -0,0 +1,140 @@
+            return
+        if not urlstr:
+            message.error(win_id, "Can't set mark with empty URL!")
+            return
```
Seeing this is called via `bookmark_add` in `commands.py`, shouldn't those checks always be true?

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -101,6 +101,11 @@ def _current_url(self):
+    def _current_title(self):
+        """Convenience method to get the current title."""
+        title = self._current_widget().title()
+        return title if title else "(null)"
```
I'm not sure if this should be handled here - maybe rather just pass the empty string to `bookmark_add`, and handle that situation there?

Also, the user interface would actually display `(null)` now, and that'd be saved to the file - why not just not save anything at all for the title, i.e. only the URL - and handle that case when reading?

The-Compiler commented on qutebrowser/completion/completiondelegate.py:
```diff
@@ -196,13 +196,10 @@ def _get_textdoc(self, index):
+            repl = r'<span class="highlight">\g<0></span>'
+            text = re.sub(re.escape(pattern), repl, self._opt.text,
+                          flags=re.IGNORECASE)
+            self._doc.setHtml(text)
```
I haven't tried yet, but I could imagine this highlighting being annoying for e.g. command completion... Maybe the completion model should be able to decide which models it wants highlighted (e.g. via a class-level constant which is used here)?

The-Compiler commented on qutebrowser/completion/models/instances.py:
```diff
@@ -108,6 +108,19 @@ def init_quickmark_completions():
+        pass
+    model = _init_model(miscmodels.BookmarkCompletionModel, 'url')
+    _instances[usertypes.Completion.bookmark_by_url] = model
+    model = _init_model(miscmodels.BookmarkCompletionModel, 'title')
```
This seems unneeded, as the "bookmark by title" model isn't used anywhere.

The-Compiler commented on qutebrowser/completion/models/miscmodels.py:
```diff
@@ -111,6 +111,27 @@ def __init__(self, match_field='url', parent=None):
                 match_field))
 
 
+class BookmarkCompletionModel(base.BaseCompletionModel):
```
Maybe this could be merged into `QuickmarkCompletionModel` as well, with a parameter (similar to `match_field` which says which object to use to get the marks (i.e. `bookmark-manager` or `quickmark-manager`)?

antoyo replied:
Removing this code duplicate will be possible if we can refactor the quickmark and bookmark managers to use a common base class.
See this comment https://github.com/The-Compiler/qutebrowser/pull/681#issuecomment-120668594.

The-Compiler commented on qutebrowser/completion/models/sortfilter.py:
```diff
@@ -137,12 +137,16 @@ def filterAcceptsRow(self, row, parent):
-            return False
-        return self.pattern.casefold() in data.casefold()
+
+        pattern = self.pattern.casefold()
```
With a custom filter, should `casefold()` really be called on the pattern already? That seems unexpected, I'd say that's rather the custom filter's responsibility if it wants to do that.

antoyo commented:
For the refactoring of bookmark and quickmark managers using a common base class, what you proposed is similar to what I tried.
The issue is that for bookmarks, the unique data (key) is the URL, while the unique data for quickmarks is the name.
So, I cannot share the code that reads the files because, with the proposed solution, I won't be able to have two bookmarks with the same description (title), unless you find a way to fix this issue.
And if we cannot share the code that reads the files between the bookmark and the quickmark managers, having a base class is useless because it will be almost empty.

antoyo commented:
Hi.
Did you get some time to check the last commits?
And could you answer to my previous comment, please?
Thanks.

The-Compiler commented:
Sorry again for the delay! I've been quite busy with stuff other than qutebrowser recently, and now at EuroPython I focused on stuff where I could speak with the maintainers of other projects when I had trouble :wink:

I've now refactored the classes so there's a common base class - this is how it looks:

```python
class UrlMarkManager(QObject):

    """Base class for BookmarkManager and QuickmarkManager.

    Attributes:
        marks: An OrderedDict of all quickmarks/bookmarks.
        _lineparser: The LineParser used for the marks, or None
                     (when qutebrowser is started with -c '').

    Signals:
        changed: Emitted when anything changed.
        added: Emitted when a new quickmark/bookmark was added.
        removed: Emitted when an existing quickmark/bookmark was removed.
    """

    changed = pyqtSignal()
    added = pyqtSignal(str, str)
    removed = pyqtSignal(str)

    def __init__(self, parent=None):
        """Initialize and read quickmarks."""
        super().__init__(parent)

        self.marks = collections.OrderedDict()
        self._lineparser = None

        if standarddir.config() is None:
            return

        self._init_lineparser()
        for line in self._lineparser:
            if not line.strip():
                # Ignore empty or whitespace-only lines.
                continue
            self._parse_line(line)
        self._init_savemanager(objreg.get('save-manager'))

    def _init_lineparser(self):
        raise NotImplementedError

    def _parse_line(self, line):
        raise NotImplementedError

    def _init_savemanager(self, _save_manager):
        raise NotImplementedError

    def save(self):
        """Save the marks to disk."""
        if self._lineparser is not None:
            self._lineparser.data = [' '.join(tpl)
                                     for tpl in self.marks.items()]
            self._lineparser.save()

    def delete(self, key):
        """Delete a quickmark/bookmark.

        Args:
            key: The key to delete (name for quickmarks, URL for bookmarks.)
        """
        del self.marks[key]
        self.changed.emit()
        self.removed.emit(key)


class QuickmarkManager(UrlMarkManager):

    """Manager for quickmarks.

    The primary key for quickmarks is their *name*, this means:

        - self.marks maps names to URLs.
        - changed gets emitted with the name as first argument and the URL as
          second argument.
        - removed gets emitted with the name as argument.
    """

    def _init_lineparser(self):
        self._lineparser = lineparser.LineParser(
            standarddir.config(), 'quickmarks', parent=self)

    def _init_savemanager(self, save_manager):
        filename = os.path.join(standarddir.config(), 'quickmarks')
        save_manager.add_saveable('quickmark-manager', self.save, self.changed,
                                  filename=filename)

    def _parse_line(self, line):
        try:
            key, url = line.rsplit(maxsplit=1)
        except ValueError:
            message.error('current', "Invalid quickmark '{}'".format(
                line))
        else:
            self.marks[key] = url

    def prompt_save(self, win_id, url):
        """Prompt for a new quickmark name to be added and add it.

        Args:
            win_id: The current window ID.
            url: The quickmark url as a QUrl.
        """
        if not url.isValid():
            urlutils.invalid_url_error(win_id, url, "save quickmark")
            return
        urlstr = url.toString(QUrl.RemovePassword | QUrl.FullyEncoded)
        message.ask_async(
            win_id, "Add quickmark:", usertypes.PromptMode.text,
            functools.partial(self.quickmark_add, win_id, urlstr))

    @cmdutils.register(instance='quickmark-manager', win_id='win_id')
    def quickmark_add(self, win_id, url, name):
        """Add a new quickmark.

        Args:
            win_id: The window ID to display the errors in.
            url: The url to add as quickmark.
            name: The name for the new quickmark.
        """
        # We don't raise cmdexc.CommandError here as this can be called async
        # via prompt_save.
        if not name:
            message.error(win_id, "Can't set mark with empty name!")
            return
        if not url:
            message.error(win_id, "Can't set mark with empty URL!")
            return

        def set_mark():
            """Really set the quickmark."""
            self.marks[name] = url
            self.changed.emit()
            self.added.emit(name, url)

        if name in self.marks:
            message.confirm_async(
                win_id, "Override existing quickmark?", set_mark, default=True)
        else:
            set_mark()

    @cmdutils.register(instance='quickmark-manager', maxsplit=0,
                       completion=[usertypes.Completion.quickmark_by_name])
    def quickmark_del(self, name):
        """Delete a quickmark.

        Args:
            name: The name of the quickmark to delete.
        """
        try:
            self.delete(name)
        except KeyError:
            raise cmdexc.CommandError("Quickmark '{}' not found!".format(name))

    def get(self, name):
        """Get the URL of the quickmark named name as a QUrl."""
        if name not in self.marks:
            raise cmdexc.CommandError(
                "Quickmark '{}' does not exist!".format(name))
        urlstr = self.marks[name]
        try:
            url = urlutils.fuzzy_url(urlstr, do_search=False)
        except urlutils.FuzzyUrlError as e:
            if e.url is None or not e.url.errorString():
                errstr = ''
            else:
                errstr = ' ({})'.format(e.url.errorString())
            raise cmdexc.CommandError("Invalid URL for quickmark {}: "
                                      "{}{}".format(name, urlstr, errstr))
        return url


class BookmarkManager(UrlMarkManager):

    """Manager for bookmarks.

    The primary key for bookmarks is their *url*, this means:

        - self.marks maps URLs to titles.
        - changed gets emitted with the URL as first argument and the title as
          second argument.
        - removed gets emitted with the URL as argument.
    """

    def _init_lineparser(self):
        bookmarks_directory = os.path.join(standarddir.config(), 'bookmarks')
        if not os.path.isdir(bookmarks_directory):
            os.makedirs(bookmarks_directory)
        self._lineparser = lineparser.LineParser(
            standarddir.config(), 'bookmarks/urls', parent=self)

    def _init_savemanager(self, save_manager):
        filename = os.path.join(standarddir.config(), 'bookmarks/urls')
        save_manager.add_saveable('bookmark-manager', self.save, self.changed,
                                  filename=filename)

    def _parse_line(self, line):
        parts = line.split(maxsplit=1)
        if len(parts) == 2:
            self.marks[parts[0]] = parts[1]
        elif len(parts) == 1:
            self.marks[parts[0]] = ''

    def add(self, win_id, url, title):
        """Add a new bookmark.

        Args:
            win_id: The window ID to display the errors in.
            url: The url to add as bookmark.
            title: The title for the new bookmark.
        """
        urlstr = url.toString(QUrl.RemovePassword | QUrl.FullyEncoded)

        if urlstr in self.marks:
            message.error(win_id, "Bookmark already exists!")
        else:
            self.marks[urlstr] = title
            self.changed.emit()
            self.added.emit(title, urlstr)
            message.info(win_id, "Bookmark added")
```

IMHO that's a lot cleaner this way :smile:

I'll leave some comments on the PR about some things I'm not sure about - please only answer the questions, I'll do the changes locally on top of my refactoring then.

After that, I think this can be merged - sorry again for the delay, I promise I'll be faster for future PRs.

The-Compiler commented on qutebrowser/browser/commands.py:
```diff
@@ -1053,6 +1057,27 @@ def quickmark_load(self, name, tab=False, bg=False, window=False):
+            bg: Load the bookmark in a new background tab.
+            window: Load the bookmark in a new window.
+        """
+        self._open(QUrl(url), tab, bg, window)
```
Why `QUrl(...)` and not `urlutils.fuzzy_url`? That's what I used for `quickmark-load`. I see both approaches could be valid... `QUrl` will potentially be less error-prone in some edge-cases, but users might also add bookmarks/quickmarks by hand maybe?

antoyo replied:
If we want to be able to add bookmarks by hand, I believe `fuzzy_url` is the way to do. I find more convenient to add bookmarks by first navigating to the page and launching the `bookmark_add` command or, if I want to import a bunch of bookmarks, by copy-pasting in the urls file directly.

The-Compiler commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -39,7 +39,11 @@ class UrlCompletionModel(base.BaseCompletionModel):
         super().__init__(parent)
 
+        self.columns_to_highlight.append(0)
+        self.columns_to_highlight.append(1)
```
Is there a reason you don't just do `self.columns_to_highlight = [0, 1]` here (and in other places)?

antoyo replied:
No, you can fix this. I think I put it this way because I was using a non-empty list in the base clase, but now the list is empty in the base class, so we can assign instead of pushing.

The-Compiler commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -81,6 +92,24 @@ def _add_quickmark_entry(self, name, url):
+        url = self.data(index0) or ''
+        title = self.data(index1) or ''
+        return (pattern.casefold() in url.casefold() or pattern.casefold() in
+                title.casefold())
```
Hmm... having custom filters is nice (thanks for adding that!) but in this case, wouldn't it make sense to rename `self.columns_to_highlight` to `columns_to_filter` or so, and implementing this in the generic filter in a generic way? It'd probably never make sense to highlight something but not filter it, right?

antoyo replied:
Yes, you're right.

The-Compiler commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -126,3 +155,50 @@ def on_quickmark_removed(self, name):
+            elif category.data() == 'Quickmarks':
+                quickmark_manager = objreg.get('quickmark-manager')
+                name = model.data(index.sibling(index.row(),
+                                  index.column() + 1))
```
Why `index.column() + 1` and not just `1` here? I'm not sure if it's always the leftmost element will be selected, but the titles will always be in column 1, right?

antoyo replied:
Yes, you're right.

The-Compiler commented on qutebrowser/completion/models/urlmodel.py:
```diff
@@ -126,3 +155,50 @@ def on_quickmark_removed(self, name):
+                name = model.data(index.sibling(index.row(),
+                                  index.column() + 1))
+                quickmark_manager.quickmark_del(name)
+                message.info(win_id, "Quickmarks deleted")
```
Do you think those messages are needed? Not having them would make this easier to test in the future, and as an user I think they are rather annoying, as you clearly can see the item is gone.

antoyo replied:
No, they are not needed.

The-Compiler commented:
Let me repeat this again :wink:

*I'll leave some comments on the PR about some things I'm not sure about - please only answer the questions, I'll do the changes locally on top of my refactoring then.*

I already did the refactoring locally (in a slightly different way than you did), so if you do the same, we'll have duplicated effort and merge conflicts :worried:

antoyo commented:
Oh sorry. I thought this was concerning the next comments. Do you want that I revert the last two commits?

The-Compiler commented:
Ah, sorry for the misunderstanding. It's fine - I'll just merge my version and then close the PR by hand - you'll still get the attribution as the original commits are by you.

The-Compiler commented:
I did a new PR at #837 with some changes on top of this. Can you please review that? :smile:

Closing this for now.
